### ISSUE TITLE
task testacularServer does nothing

### ISSUE THREAD

**Issue by marcorinck** [dignifiedquire/grunt-testacular#5]
I'm quite new to node.js, grunt and even testacular. So please bare with me, if I'm doing a stupid mistake. 

I wanted to use testacular tests in my first grunt project. I'm currently using testacular 0.5.5 grunt-testacular 0.3.0 and node.js 0.8.14 on WIndows 7.

I have a testacular.conf.js which is working when run with testacular directly via "testacular start". When I then use "grunt testacularRun" its running the tests with the expected results (its reporting failed tests). 

But when I run "grunt testacularServer" its doing nothing. I get this output: 

```
C:\Users\mzrinck\Documents\grunt\test1>grunt.cmd testacularServer
Running "testacularServer:unit" (testacularServer) task

Done, without errors.
```

This is my task in grunt.js:

``` javascript
testacularServer: {
    unit: {
        configFile: 'testacular.conf.js'
    }
},
```

The grunt.js and testacular.conf.js are in the same folder. 

What is wrong here?

Thanks for any help.

**Comment by dignifiedquire** [comment]
Hey the thing is that if you want to use the `testacularServer` taks without any other grunt tasks you need to add the option `keepalive: true` to keep grunt running as long as testacular is running.  So your config file woul look like this

``` js
testacularServer: {
  unit: {
    options: {
      keepalive: true
    },
    configFile: 'config/testacular.conf.js'
  }
}
```

**Comment by marcorinck** [comment]
Thanks for your help. 

If I set keepAlive true, testacularServer runs and keeps watching the tests. But it never ends. 

But what I want to achieve is a single run of unit tests during build and deployment. Testacular itself in latest version has a "singleRun" option, which just does that (for continious integration purpose). 

Can I achieve something like that with grunt-testacular, too?

**Comment by dignifiedquire** [comment]
Yes you have two options either you set singleRun: true in your gruntfile or in testeculars config file singleRun=true

**Comment by marcorinck** [comment]
yes, I have found that myself now in the closed issues. There was an example. Closing it now. 

Thanks!

**Comment by marcorinck** [comment]
sorry, to disturb you again :-(

But I haven't understood it completely as I have now the same problem as in isuee #3. 

When I set keepAlive and singleRun to true, the testacularServer task is executed and grunt exits afterwards when its the only task or the last task in chain. But when I want to run tasks after testacularServer they are not executed at all. 

When I set singleRun to true and keepAlive to false, all tasks are executed but testacularServer does nothing at all (besides a message that nothing was done). 

When doing the build I want to execute my tests once in the beginning and only when all tests are green the real build tasks should be executed. How do I do this?

**Comment by dignifiedquire** [comment]
I'm afraid that is not easily possible right now. As you are on windows it gets even more complicated. Testacular supports different exit codes depending on the tests passing or not but you would have to implement your own script to listen to this event as grunt has no way right now of calling another task depending on the previous task.

**Comment by marcorinck** [comment]
Would that be possible on Linux?

**Comment by dignifiedquire** [comment]
I've thought about it and maybe it's even possible to implement cross platform. I'll have a look in the next days and let you if I come up with something.

**Comment by marcorinck** [comment]
Awesome, thanks!

**Comment by dignifiedquire** [comment]
I've got it working and as soon as the above mentioned PR gets merged I'll publish it.
If you have a look at the readme of the canary branch you can see how it's going to work.

**Comment by marcorinck** [comment]
I have merged your changes in my local testacular installation manually (0.5.5, grunt 0.4.0a). 

After that the task "testacularServer" doesn't end at all even if its the only task for grunt (when keepalive is true, when is false, it still does nothing) 

```
Running "testacularServer:unit" (testacularServer) task
info: Testacular server started at http://localhost:10000/
info (launcher): Starting browser Chrome
debug (launcher): Creating temp dir at C:\Users\mzrinck\AppData\Local\Temp\testacular-53091185
debug (launcher): C:\Users\mzrinck\AppData\Local\Google\Chrome\Application\chrome.exe --user-data-dir=C:\Users\mzrinck\AppData\Local\Temp\testacular-5
3091185 --no-default-browser-check --no-first-run --disable-default-apps http://localhost:10000/?id=53091185
debug (watcher): Resolved files:
        c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/adapter/lib/jasmine.js
        c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/adapter/jasmine.js
        c:/Users/mzrinck/Documents/grunt/hwa-mobile/src/test/testSpec.js
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/static/client.html
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/static/testacular.js
debug: New browser has connected on socket WzVjKznsMTdQPTB1Lm-F
info (Chrome 23.0 (Windows)): Connected on socket id WzVjKznsMTdQPTB1Lm-F
debug: All browsers are ready, executing
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/static/context.html
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/src/test/testSpec.js
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/adapter/jasmine.js
debug (web server): serving: c:/Users/mzrinck/Documents/grunt/hwa-mobile/node_modules/grunt-testacular/node_modules/testacular/adapter/lib/jasmine.js
Chrome 23.0 (Windows): Executed 1 of 1 SUCCESS (0.062 secs / 0.002 secs)
debug (launcher): Disconnecting all browsers
debug (launcher): Killing Chrome
debug (launcher): Process Chrome exitted with code 0
debug (launcher): Cleaning temp dir C:\Users\mzrinck\AppData\Local\Temp\testacular-53091185
```

**Comment by dignifiedquire** [comment]
Are you using singleRun?Otherwise you need to use the testacularRun task

**Comment by marcorinck** [comment]
yes, singleRun is true:

```
testacularServer: {
            unit: {
                configFile: 'testacular.conf.js',
                options: {
                    keepalive: false
                },
                autoWatch: false,
                singleRun: true
            }

        },
```

**Comment by dignifiedquire** [comment]
Ok if you want to use the `testacularServer` task with single run you need to do something like this

``` js
testacularServer: {
            unit: {
                configFile: 'testacular.conf.js',
                options: {
                    keepalive: true,
                    success: 'successTask',
                    failure: 'failureTask'
                },
                autoWatch: false,
                singleRun: true
            }

        },
```

With this configuration it would run your tests once and then execute the `successTask` if the tasks passed and the `failureTask` if the tests failed.

**Comment by mlegenhausen** [comment]
Would it be possible to have the same execution behaviour as other tasks? Cause it is not practical to define the `success` and `failure` task. When you have multiple tasks defined via `registerTask` you need to define an own configuration for each task. It would be nice if testacular would simply stop the task execution on failure and continue task execution when all tests succeed.

**Comment by dignifiedquire** [comment]
Yes that should be possible. I'll push something tonight.

**Comment by jaketrent** [comment]
+1  This is getting better and better.

**Comment by dignifiedquire** [comment]
Thanks. I will finish this as soon as the stable release 0.6 is out. I have
found a nice api that will make this solve this problem very nicely. Stay
stuned :)

**Comment by jaketrent** [comment]
@Dignifiedquire I'm looking at the new syntax and wondering why you would need a `success` and `failure` callback specified.

With your merged pull request [206](https://github.com/vojtajina/testacular/pull/206) into testacular, now your plugin would be able to detect test failure/passing and call grunt.async(false/true).

If I modify `testacularServer.js` to do this much:

``` js
var done = this.async();
server.start(this.data, function(exitCode) {
   return done(exitCode == 0);
});
```

Set my grunt config:

``` js
testacularServer: {
  ci: {
    configFile: 'testacular.conf.js',
    options: {
      keepalive: true
    },
    autoWatch: false,
    singleRun: true
  }
}
```

And run grunt:

```
grunt testacularServer:ci lint
```

If the testacular tests pass, lint runs;  when the tests fail, lint does not run.  (lint could be any other task.)  Couldn't we modify testacularServer.js in this small way and get what we need: Tasks after testacular don't run unless tests pass?

**Comment by dignifiedquire** [comment]
No that's the old new syntax ;)
My new new syntax can't be seen anywhere as I haven't written in done anywhere. But it's basically what you suggest only that I'll use grunts warn functionality to cancel the task, so that one has the ability to continue running when used with the `--force` flag.

**Comment by dignifiedquire** [comment]
@jaketrent With the new version `0.3.0` and your config everything should be good to go. Please let me know if it works for you.
